src/layers/core/arc-layer/arc-layer-vertex.glsl

```diff
@@ -43,22 +44,41 @@ float paraboloid(vec2 source, vec2 target, float ratio) {
   return (dSourceCenter + dXCenter) * (dSourceCenter - dXCenter);
 }
 
-void main(void) {
-  vec2 source = preproject(instancePositions.xy);
-  vec2 target = preproject(instancePositions.zw);
-
-  float segmentRatio = smoothstep(0.0, 1.0, positions.x / N);
+float getSegmentRatio(float index) {
+  return smoothstep(0.0, 1.0, index / (numSegments - 1.0));
+}
 
+vec3 getPos(vec2 source, vec2 target, float segmentRatio) {
   float vertex_height = paraboloid(source, target, segmentRatio);
-  if (vertex_height < 0.0) vertex_height = 0.0;
-  vec3 p = vec3(
-    // xy: linear interpolation of source & target
+
+  return vec3(
     mix(source, target, segmentRatio),
-    // z: paraboloid interpolate of source & target
-    sqrt(vertex_height)
+    sqrt(max(0.0, vertex_height))
   );
+}
+
+void main(void) {
+  vec2 source = project_position(instancePositions.xy);
+  vec2 target = project_position(instancePositions.zw);
+
+  float segmentIndex = positions.x;
+  float segmentRatio = getSegmentRatio(segmentIndex);
+  // if it's the first point, use next - current as direction
+  // otherwise use current - prev
+  float indexDir = mix(-1.0, 1.0, step(segmentIndex, 0.0));
+  float nextSegmentRatio = getSegmentRatio(segmentIndex + indexDir);
+  
+  vec3 currPos = getPos(source, target, segmentRatio);
+  vec3 nextPos = getPos(source, target, nextSegmentRatio);
+  vec4 curr = project_to_clipspace(vec4(currPos, 1.0));
+  vec4 next = project_to_clipspace(vec4(nextPos, 1.0));
+
+  // extrude by strokeWidth
+  vec2 dir = normalize((next.xy - curr.xy) * indexDir * screenSize);
+  vec2 perp = vec2(-dir.y, dir.x);
+  vec2 offset = perp * positions.y * strokeWidth / screenSize;
```

Comment:
Might be nice to define a small well-named and well-commented function that does the move in pixel space.